packages/compass-explain-plan/src/components/explain-tree/explain-tree.tsx

```diff
@@ -0,0 +1,92 @@
+import React, { useRef, useEffect } from 'react';
+import { map } from 'lodash';
+import d3 from 'd3';
+import { ExplainStage } from '../explain-stage';
+import {
+  css,
+  palette,
+  spacing,
+  useDarkMode,
+} from '@mongodb-js/compass-components';
+
+import STAGE_CARD_PROPERTIES from '../../constants/stage-card-properties';
+
+interface ExplainTreeProps {
+  nodes: any[];
+  links: any[];
+  width: number;
+  height: number;
+}
+
+const explainTreeStyles = css({
+  position: 'relative',
+  zIndex: 0,
+  margin: `${spacing[5]}px auto`,
+});
+
+const ExplainTree: React.FunctionComponent<ExplainTreeProps> = ({
+  nodes,
+  links,
+  width,
+  height,
+}) => {
+  const svgRef = useRef<SVGSVGElement>(null);
+  const darkMode = useDarkMode();
+
+  useEffect(() => {
+    if (!svgRef.current || !links || !nodes) return;
+    const svgElement = svgRef.current;
+
+    // Right angle links between nodes
+    const elbow = (d: {
+      source: { x: number; y: number; x_size: number };
+      target: { y: number; x: number; key: string };
+    }): string => {
+      return `M${d.source.x + d.source.x_size / 2},${d.source.y}
+        V${d.target.y - STAGE_CARD_PROPERTIES.VERTICAL_PADDING / 2}
+        H${d.target.x + STAGE_CARD_PROPERTIES.DEFAULT_CARD_WIDTH / 2}
+        V${d.target.y}`;
+    };
+    const svg = d3.select(svgElement);
+
+    // Links are svg elements
+    const svgLinks = svg
+      .selectAll('path')
+      .data(links, (d) => d.target.key)
+      .attr('d', elbow);
+
+    svgLinks
+      .enter()
+      .append('path')
+      .style({
+        fill: 'none',
+        stroke: darkMode ? palette.gray.base : palette.gray.light2,
+        'stroke-width': '6px',
+      })
+      .attr('d', elbow);
+
+    svgLinks.exit().remove();
+
+    return () => {
+      // make sure to clean up the elements from the previous hook execution
+      d3.select(svgElement).selectAll('*').remove();
+    };
+  }, [nodes, links, darkMode]);
+
+  const getStages = () => {
+    return map(nodes, (stage) => <ExplainStage {...stage} />);
+  };
```

Comment:
Feels a bit weird to me that here, and only here, we are wrapping this in a method and use lodash to create an array instead of just doing a map like everywhere else. You also need a key properties when you are rendering arrays of elements in React

Reply:
same stuff, i basically rewrote this file, moved the layout part from the store to the component and have 2 components a TreeLayout and a ExplainTree:

```
<TreeLayout<ExplainTreeNodeData>
        data-testid="explain-tree"
        data={root}
        getNodeSize={getNodeSize}
        getNodeKey={getNodeKey}
        linkColor={darkMode ? palette.gray.base : palette.gray.light2}
        linkWidth={6}
        horizontalSpacing={spacing[6]}
        verticalSpacing={spacing[6]}
        className={explainTreeStyles}
      >
        {(node) => {
          const key = getNodeKey(node);
          return (
            <div
              style={{
                position: 'relative',
                zIndex: detailsOpen === key ? 2 : 1,
              }}
            >
              <ExplainStage
                detailsOpen={detailsOpen === key}
                toggleDetails={() => {
                  detailsOpen === key
                    ? setDetailsOpen(null)
                    : setDetailsOpen(key);
                }}
                {...node}
                totalExecTimeMS={root.curStageExecTimeMS}
              ></ExplainStage>
            </div>
          );
        }}
      </TreeLayout>
```

I've also finally understood what is the rule behind drawing the links, we could get rid of d3 there as you suggested, not sure how much we should improve that layout though, I did an attempt with aggregations and the tree is gigantic to the point that we probably need some zoom + pan feature

Reply:
Oh, cool, looking forward to that patch!